File: multi_language/session_raid_stats/session_raid_stats.R
# session_raid_stats.R — R port of the Session Raid Stats Pine Script indicator
# ==============================================================================
# Tracks up to three intraday sessions, detects range-extension raids, buckets
# raid sizes, and computes empirical reach-probabilities for each level.
#
# Dependencies: base R only (no external packages required).
#
# Usage:
#   source("session_raid_stats.R")
#   cfg <- session_config("02:00", "02:15", min_raid_pts = 5.0, cutoff_mins = 120,
#                          bucket_start = 20.0, bucket_step = 10.0)
#   engine <- raid_engine(cfg)
#   for (i in seq_len(nrow(bars))) {
#     engine$on_bar(bars$time_ms[i], bars$open[i], bars$high[i], bars$low[i], bars$close[i])
#   }
#   stats <- engine$get_stats()
#   print(stats$prob_hi)

MAX_BUCKETS <- 6L
BC          <- MAX_BUCKETS + 1L   # 7 buckets: 0..5 + overflow

# ---------------------------------------------------------------------------
# Configuration
# ---------------------------------------------------------------------------

session_config <- function(session_start, session_end,
                            min_raid_pts  = 5.0,
                            cutoff_mins   = 120L,
                            bucket_start  = 20.0,
                            bucket_step   = 10.0,
                            tz_offset_hrs = -5.0) {
  level_pts <- function(n) bucket_start + (n - 1L) * bucket_step
  list(
    session_start  = session_start,
    session_end    = session_end,
    min_raid_pts   = min_raid_pts,
    cutoff_mins    = as.integer(cutoff_mins),
    bucket_start   = bucket_start,
    bucket_step    = bucket_step,
    tz_offset_hrs  = tz_offset_hrs,
    levels         = sapply(1:MAX_BUCKETS, level_pts)
  )
}

# ---------------------------------------------------------------------------
# Bucket helpers
# ---------------------------------------------------------------------------

bucket_index <- function(value, cfg) {
  lvls <- cfg$levels
  for (i in seq_len(MAX_BUCKETS - 1L)) {
    if (value < lvls[i + 1L]) return(i - 1L)   # 0-based
  }
  if (value < lvls[MAX_BUCKETS] + cfg$bucket_step) return(MAX_BUCKETS - 1L)
  return(MAX_BUCKETS)   # overflow
}

cumulative_counts <- function(counts) {
  n   <- length(counts)
  cum <- integer(n)
  running <- 0L
  for (i in n:1L) {
    running <- running + counts[i]
    cum[i]  <- running
  }
  cum
}

update_prob_cache <- function(hi_counts, lo_counts, day_count, cache_hi, cache_lo) {
  if (day_count <= 0L) return(list(hi = cache_hi, lo = cache_lo))
  cum_h <- cumulative_counts(hi_counts)
  cum_l <- cumulative_counts(lo_counts)
  list(
    hi = cum_h / day_count * 100.0,
    lo = cum_l / day_count * 100.0
  )
}

# ---------------------------------------------------------------------------
# Session time helpers
# ---------------------------------------------------------------------------

.parse_hhmm <- function(s) {
  parts <- strsplit(s, ":")[[1L]]
  list(h = as.integer(parts[1L]), m = as.integer(parts[2L]))
}

.bar_local_time <- function(time_ms, tz_offset_hrs) {
  utc_sec   <- time_ms / 1000.0
  local_sec <- utc_sec + tz_offset_hrs * 3600.0
  as.POSIXct(local_sec, origin = "1970-01-01", tz = "UTC")
}

.in_session <- function(time_ms, cfg) {
  dt   <- .bar_local_time(time_ms, cfg$tz_offset_hrs)
  h    <- as.integer(format(dt, "%H"))
  m    <- as.integer(format(dt, "%M"))
  mins <- h * 60L + m
  s    <- .parse_hhmm(cfg$session_start)
  e    <- .parse_hhmm(cfg$session_end)
  s_m  <- s$h * 60L + s$m
  e_m  <- e$h * 60L + e$m
  mins >= s_m && mins < e_m
}

.within_cutoff <- function(time_ms, range_end_ms, cutoff_mins) {
  if (cutoff_mins == 0L) return(TRUE)
  (time_ms - range_end_ms) <= cutoff_mins * 60000L
}

# ---------------------------------------------------------------------------
# Engine factory (returns a closure with mutable state)
# ---------------------------------------------------------------------------

raid_engine <- function(cfg, max_days = 500L) {
  # Mutable state stored in the closure environment
  st <- new.env(parent = emptyenv())
  st$hi        <- NA_real_
  st$lo        <- NA_real_
  st$end_ms    <- NA_integer_
  st$active    <- FALSE
  st$hi_max    <- NA_real_
  st$hi_touch  <- FALSE
  st$hi_conf   <- FALSE
  st$hi_pts    <- NA_real_
  st$lo_min    <- NA_real_
  st$lo_touch  <- FALSE
  st$lo_conf   <- FALSE
  st$lo_pts    <- NA_real_
  st$cut_exp   <- FALSE

  hi_counts    <- integer(BC)
  lo_counts    <- integer(BC)
  prob_hi      <- double(BC)
  prob_lo      <- double(BC)
  day_count    <- 0L
  prev_in_s    <- FALSE

  on_bar <- function(time_ms, open_, high, low, close) {
    in_s <- .in_session(time_ms, cfg)

    # Session open — reset state
    if (in_s && !prev_in_s) {
      st$hi      <<- high;  st$lo <- low
      st$end_ms  <<- NA;    st$active <- FALSE
      st$hi_max  <<- NA;    st$hi_touch <- FALSE
      st$hi_conf <<- FALSE; st$hi_pts <- NA
      st$lo_min  <<- NA;    st$lo_touch <- FALSE
      st$lo_conf <<- FALSE; st$lo_pts <- NA
      st$cut_exp <<- FALSE
    }

    # Expand range during session
    if (in_s && !is.na(st$hi)) {
      st$hi <<- max(st$hi, high)
      st$lo <<- min(st$lo, low)
    }

    # Session close — activate raid detection
    if (!in_s && prev_in_s && !is.na(st$hi)) {
      st$end_ms <<- time_ms
      st$active <<- TRUE
      if (day_count < max_days) day_count <<- day_count + 1L
    }

    # Raid detection
    if (st$active && !is.na(st$end_ms)) {
      if (.within_cutoff(time_ms, st$end_ms, cfg$cutoff_mins)) {
        # High-side
        if (high > st$hi) {
          ext <- high - st$hi
          if (ext >= cfg$min_raid_pts) {
            if (is.na(st$hi_max) || high > st$hi_max) {
              st$hi_max   <<- high
              st$hi_pts   <<- ext
              st$hi_touch <<- TRUE
            }
          }
        }
        # Low-side
        if (low < st$lo) {
          ext <- st$lo - low
          if (ext >= cfg$min_raid_pts) {
            if (is.na(st$lo_min) || low < st$lo_min) {
              st$lo_min   <<- low
              st$lo_pts   <<- ext
              st$lo_touch <<- TRUE
            }
          }
        }
      } else if (!st$cut_exp) {
        if (st$hi_touch && !is.na(st$hi_pts)) {
          idx <- bucket_index(st$hi_pts, cfg) + 1L   # 1-based for R
          hi_counts[idx] <<- hi_counts[idx] + 1L
          st$hi_conf <<- TRUE
        }
        if (st$lo_touch && !is.na(st$lo_pts)) {
          idx <- bucket_index(st$lo_pts, cfg) + 1L
          lo_counts[idx] <<- lo_counts[idx] + 1L
          st$lo_conf <<- TRUE
        }
        cache <- update_prob_cache(hi_counts, lo_counts, day_count, prob_hi, prob_lo)
        prob_hi <<- cache$hi
        prob_lo <<- cache$lo
        st$cut_exp <<- TRUE
      }
    }
    prev_in_s <<- in_s
  }

  get_stats <- function() {
    list(prob_hi = prob_hi, prob_lo = prob_lo, day_count = day_count)
  }

  list(on_bar = on_bar, get_stats = get_stats)
}

# ---------------------------------------------------------------------------
# Three-session facade
# ---------------------------------------------------------------------------

session_raid_stats <- function(cfg1, cfg2, cfg3, max_days = 500L) {
  e1 <- raid_engine(cfg1, max_days)
  e2 <- raid_engine(cfg2, max_days)
  e3 <- raid_engine(cfg3, max_days)

  list(
    on_bar = function(t, o, h, l, c) {
      e1$on_bar(t, o, h, l, c)
      e2$on_bar(t, o, h, l, c)
      e3$on_bar(t, o, h, l, c)
    },
    stats = function() list(r1 = e1$get_stats(), r2 = e2$get_stats(), r3 = e3$get_stats())
  )
}

# ---------------------------------------------------------------------------
# Quick self-test
# ---------------------------------------------------------------------------

if (!interactive()) {
  set.seed(42L)
  cfg <- session_config("02:00", "02:15", min_raid_pts = 5.0, cutoff_mins = 120L,
                         bucket_start = 20.0, bucket_step = 10.0)
  eng <- raid_engine(cfg, max_days = 100L)

  # 2024-01-02 07:00 UTC = 02:00 ET
  base_ms  <- 1704182400000
  ms_min   <- 60000L
  price    <- 4500.0

  for (day in 0:4) {
    day_off <- day * 24L * 60L * ms_min
    for (minute in 0:(8 * 60 - 1)) {
      t  <- base_ms + day_off + minute * ms_min
      o  <- price
      h  <- o + runif(1, 0, 4)
      l  <- o - runif(1, 0, 4)
      c  <- runif(1, l, h)
      price <- c
      eng$on_bar(t, o, h, l, c)
    }
  }

  s <- eng$get_stats()
  cat("Day count:", s$day_count, "\n")
  cat("prob_hi  :", paste0(round(s$prob_hi, 1), "%", collapse = " "), "\n")
  cat("prob_lo  :", paste0(round(s$prob_lo, 1), "%", collapse = " "), "\n")
}

File: multi_language/session_raid_stats/session_raid_stats.py
"""
session_raid_stats.py — Python 3 port of the Session Raid Stats Pine Script indicator
======================================================================================
Tracks up to three intraday sessions, detects range-extension raids, buckets raid
sizes, and computes empirical reach-probabilities for each level.

Usage
-----
    from session_raid_stats import SessionConfig, RaidEngine, RaidStats

    cfg = SessionConfig(
        session_start="02:00", session_end="02:15",
        min_raid_pts=5.0, cutoff_mins=120,
        bucket_start=20.0, bucket_step=10.0,
    )
    engine = RaidEngine(cfg)

    # Feed OHLC bars in chronological order:
    for bar in ohlc_bars:
        engine.on_bar(bar["time_ms"], bar["open"], bar["high"], bar["low"], bar["close"])

    stats = engine.get_stats()
    print(stats.prob_hi)  # cumulative reach-probability per level (high-side raids)
    print(stats.prob_lo)  # cumulative reach-probability per level (low-side raids)
"""

from __future__ import annotations

import math
from dataclasses import dataclass, field
from datetime import datetime, timezone, time as dtime
from typing import List, Optional


# ---------------------------------------------------------------------------
# Configuration
# ---------------------------------------------------------------------------

MAX_BUCKETS = 6
BC = MAX_BUCKETS + 1  # 7 buckets: levels 0-5 + overflow


@dataclass
class SessionConfig:
    session_start: str   # "HH:MM"  (local exchange time, 24-h)
    session_end:   str   # "HH:MM"
    min_raid_pts:  float = 5.0
    cutoff_mins:   int   = 120    # 0 = no cutoff
    bucket_start:  float = 20.0
    bucket_step:   float = 10.0
    tz_offset_hrs: float = -5.0   # ET = UTC-5 (no DST adjustment in this simplified port)

    def level_pts(self, n: int) -> float:
        """Return the threshold in points for level n (1-indexed)."""
        return self.bucket_start + (n - 1) * self.bucket_step

    @property
    def levels(self) -> List[float]:
        return [self.level_pts(n) for n in range(1, MAX_BUCKETS + 1)]


# ---------------------------------------------------------------------------
# Bucket helpers
# ---------------------------------------------------------------------------

def bucket_index(value: float, cfg: SessionConfig) -> int:
    """Map a raid-size value to a 0-based bucket index (overflow = MAX_BUCKETS)."""
    lvls = cfg.levels
    for i in range(MAX_BUCKETS - 1):
        if value < lvls[i + 1]:
            return i
    # Check if it fits in the last named bucket
    if value < lvls[-1] + cfg.bucket_step:
        return MAX_BUCKETS - 1
    return MAX_BUCKETS  # overflow


def cumulative_counts(counts: List[int]) -> List[int]:
    """Reverse-cumulative sum: cum[i] = sum(counts[i:])."""
    cum = [0] * len(counts)
    running = 0
    for i in range(len(counts) - 1, -1, -1):
        running += counts[i]
        cum[i] = running
    return cum


def update_prob_cache(
    hi_counts: List[int],
    lo_counts: List[int],
    day_count: int,
    cache_hi: List[float],
    cache_lo: List[float],
) -> None:
    """Update ECDF probability caches in-place."""
    if day_count <= 0:
        return
    cum_h = cumulative_counts(hi_counts)
    cum_l = cumulative_counts(lo_counts)
    for i in range(len(cache_hi)):
        cache_hi[i] = cum_h[i] / day_count * 100.0
        cache_lo[i] = cum_l[i] / day_count * 100.0


# ---------------------------------------------------------------------------
# Session time helpers
# ---------------------------------------------------------------------------

def _parse_hhmm(s: str):
    parts = s.split(":")
    return int(parts[0]), int(parts[1])


def _bar_local_time(bar_time_ms: int, tz_offset_hrs: float) -> dtime:
    utc_ts = bar_time_ms / 1000.0
    offset_secs = tz_offset_hrs * 3600
    local_ts = utc_ts + offset_secs
    dt = datetime.utcfromtimestamp(local_ts)
    return dt.time()


def _in_session(bar_time_ms: int, cfg: SessionConfig) -> bool:
    t = _bar_local_time(bar_time_ms, cfg.tz_offset_hrs)
    sh, sm = _parse_hhmm(cfg.session_start)
    eh, em = _parse_hhmm(cfg.session_end)
    start = dtime(sh, sm)
    end   = dtime(eh, em)
    return start <= t < end


def _within_cutoff(bar_time_ms: int, range_end_ms: int, cutoff_mins: int) -> bool:
    if cutoff_mins == 0:
        return True
    return (bar_time_ms - range_end_ms) <= cutoff_mins * 60_000


# ---------------------------------------------------------------------------
# Per-session raid state
# ---------------------------------------------------------------------------

@dataclass
class RaidState:
    hi:        Optional[float] = None
    lo:        Optional[float] = None
    end_ms:    Optional[int]   = None
    active:    bool            = False
    hi_max:    Optional[float] = None
    hi_touch:  bool            = False
    hi_conf:   bool            = False
    hi_pts:    Optional[float] = None
    lo_min:    Optional[float] = None
    lo_touch:  bool            = False
    lo_conf:   bool            = False
    lo_pts:    Optional[float] = None
    cut_exp:   bool            = False


@dataclass
class RaidStats:
    prob_hi: List[float] = field(default_factory=lambda: [0.0] * BC)
    prob_lo: List[float] = field(default_factory=lambda: [0.0] * BC)
    day_count: int = 0


# ---------------------------------------------------------------------------
# Engine
# ---------------------------------------------------------------------------

class RaidEngine:
    """Processes chronological OHLC bars for one session configuration."""

    def __init__(self, cfg: SessionConfig, max_days: int = 500):
        self.cfg       = cfg
        self.max_days  = max_days
        self._state    = RaidState()
        self._hi_counts: List[int]   = [0] * BC
        self._lo_counts: List[int]   = [0] * BC
        self._prob_hi:   List[float] = [0.0] * BC
        self._prob_lo:   List[float] = [0.0] * BC
        self._day_count  = 0
        self._prev_in_session = False

    # ------------------------------------------------------------------
    def on_bar(self, time_ms: int, open_: float, high: float, low: float, close: float) -> None:
        cfg   = self.cfg
        state = self._state
        in_s  = _in_session(time_ms, cfg)

        # --- Session open: capture first bar
        if in_s and not self._prev_in_session:
            state.hi      = high
            state.lo      = low
            state.end_ms  = None
            state.active  = False
            state.hi_max  = None
            state.hi_touch = False
            state.hi_conf  = False
            state.hi_pts   = None
            state.lo_min   = None
            state.lo_touch = False
            state.lo_conf  = False
            state.lo_pts   = None
            state.cut_exp  = False

        # --- Session body: expand range
        if in_s and state.hi is not None:
            state.hi = max(state.hi, high)
            state.lo = min(state.lo, low)

        # --- Session close: activate raid detection
        if not in_s and self._prev_in_session and state.hi is not None:
            state.end_ms = time_ms
            state.active = True
            if self._day_count < self.max_days:
                self._day_count += 1

        # --- Raid detection (post-session)
        if state.active and state.end_ms is not None:
            if _within_cutoff(time_ms, state.end_ms, cfg.cutoff_mins):
                min_pts = cfg.min_raid_pts
                # High-side raid
                if high > state.hi:
                    ext = high - state.hi
                    if ext >= min_pts:
                        if state.hi_max is None or high > state.hi_max:
                            state.hi_max  = high
                            state.hi_pts  = ext
                            state.hi_touch = True
                # Low-side raid
                if low < state.lo:
                    ext = state.lo - low
                    if ext >= min_pts:
                        if state.lo_min is None or low < state.lo_min:
                            state.lo_min  = low
                            state.lo_pts  = ext
                            state.lo_touch = True
            else:
                # Cutoff expired — lock in confirmed raids for this session
                if not state.cut_exp:
                    if state.hi_touch and state.hi_pts is not None:
                        idx = bucket_index(state.hi_pts, cfg)
                        self._hi_counts[idx] += 1
                        state.hi_conf = True
                    if state.lo_touch and state.lo_pts is not None:
                        idx = bucket_index(state.lo_pts, cfg)
                        self._lo_counts[idx] += 1
                        state.lo_conf = True
                    update_prob_cache(
                        self._hi_counts, self._lo_counts, self._day_count,
                        self._prob_hi, self._prob_lo,
                    )
                    state.cut_exp = True

        self._prev_in_session = in_s

    # ------------------------------------------------------------------
    def get_stats(self) -> RaidStats:
        return RaidStats(
            prob_hi=self._prob_hi[:],
            prob_lo=self._prob_lo[:],
            day_count=self._day_count,
        )


# ---------------------------------------------------------------------------
# Three-session facade (mirrors the Pine Script indicator)
# ---------------------------------------------------------------------------

class SessionRaidStats:
    """Manages three raid engines simultaneously."""

    def __init__(
        self,
        cfg1: SessionConfig,
        cfg2: SessionConfig,
        cfg3: SessionConfig,
        max_days: int = 500,
    ):
        self.engine1 = RaidEngine(cfg1, max_days)
        self.engine2 = RaidEngine(cfg2, max_days)
        self.engine3 = RaidEngine(cfg3, max_days)

    def on_bar(self, time_ms: int, open_: float, high: float, low: float, close: float) -> None:
        self.engine1.on_bar(time_ms, open_, high, low, close)
        self.engine2.on_bar(time_ms, open_, high, low, close)
        self.engine3.on_bar(time_ms, open_, high, low, close)

    def stats(self):
        return (
            self.engine1.get_stats(),
            self.engine2.get_stats(),
            self.engine3.get_stats(),
        )


# ---------------------------------------------------------------------------
# Quick self-test
# ---------------------------------------------------------------------------

if __name__ == "__main__":
    import random

    random.seed(42)
    cfg = SessionConfig("02:00", "02:15", min_raid_pts=5.0, cutoff_mins=120,
                        bucket_start=20.0, bucket_step=10.0)
    engine = RaidEngine(cfg, max_days=100)

    # Synthetic 1-minute bars for 5 trading days starting 2024-01-02 07:00 UTC (= 02:00 ET)
    base_price = 4500.0
    ms_per_min = 60_000
    session_open_utc_ms = 1704182400_000  # 2024-01-02 07:00 UTC

    for day in range(5):
        day_offset = day * 24 * 60 * ms_per_min
        price = base_price
        for minute in range(8 * 60):  # 8-hour window
            t = session_open_utc_ms + day_offset + minute * ms_per_min
            o = price
            h = o + random.uniform(0, 4)
            l = o - random.uniform(0, 4)
            c = random.uniform(l, h)
            price = c
            engine.on_bar(t, o, h, l, c)

    stats = engine.get_stats()
    print("Day count :", stats.day_count)
    print("prob_hi   :", [f"{p:.1f}%" for p in stats.prob_hi])
    print("prob_lo   :", [f"{p:.1f}%" for p in stats.prob_lo])
